src/llmq/quorums_chainlocks.cpp

```diff
@@ -0,0 +1,350 @@
+        tmp.blockHash = pindexNew->GetBlockHash();
+        if (!myVotes.emplace(requestId, tmp).second) {
+            // already votes on this one
+            return;
+        }
+    }
+
+    quorumSigningManager->AsyncSignIfMember(Params().GetConsensus().llmqTypeForChainLocks, requestId, msgHash);
+}
+
+void CChainLocksHandler::HandleNewRecoveredSig(const llmq::CRecoveredSig& recoveredSig)
+{
+    if (!sporkManager.IsSporkActive(SPORK_19_CHAINLOCKS_ENABLED)) {
+        return;
+    }
+
+    CChainLockSig clsig;
+    {
+        LOCK(cs);
+
+        auto it = myVotes.find(recoveredSig.id);
+        if (it == myVotes.end()) {
+            return;
+        }
+        if (it->second.sig.IsValid()) {
+            // already got a valid sig
+            return;
+        }
+        if (it->second.blockHash != recoveredSig.msgHash) {
+            // does not match out vote
+            return;
+        }
+
+        it->second.sig = recoveredSig.sig;
+        clsig = it->second;
+    }
+    ProcessNewChainLock(-1, clsig, ::SerializeHash(clsig));
+}
+
+void CChainLocksHandler::ScheduleInvalidateBlock(const CBlockIndex* pindex)
+{
+    // Calls to InvalidateBlock and ActivateBestChain might result in re-invocation of the UpdatedBlockTip and other
+    // signals, so we can't directly call it from signal handlers. We solve this by doing the call from the scheduler
+
+    scheduler->scheduleFromNow([this, pindex]() {
+        DoInvalidateBlock(pindex, true);
+    }, 0);
+}
+
+void CChainLocksHandler::DoInvalidateBlock(const CBlockIndex* pindex, bool activateBestChain)
+{
+    auto& params = Params();
+
+    {
+        LOCK(cs_main);
+
+        // get the non-const pointer
+        CBlockIndex* pindex2 = mapBlockIndex[pindex->GetBlockHash()];
+
+        CValidationState state;
+        if (!InvalidateBlock(state, params, pindex2)) {
+            LogPrintf("CChainLocksHandler::UpdatedBlockTip -- InvalidateBlock failed: %s\n", FormatStateMessage(state));
+            // This should not have happened and we are in a state were it's not safe to continue anymore
+            assert(false);
+            return;
+        }
+    }
+
+    CValidationState state;
+    if (activateBestChain && !ActivateBestChain(state, params)) {
+        LogPrintf("CChainLocksHandler::UpdatedBlockTip -- ActivateBestChain failed: %s\n", FormatStateMessage(state));
+        // This should not have happened and we are in a state were it's not safe to continue anymore
+        assert(false);
+        return;
+    }
+}
+
+bool CChainLocksHandler::HasChainLock(const CBlockIndex* pindexPrev, const uint256& newBlockHash)
+{
+    if (!sporkManager.IsSporkActive(SPORK_19_CHAINLOCKS_ENABLED)) {
+        return false;
+    }
+
+    LOCK2(cs_main, cs);
+    return InternalHasChainLock(pindexPrev, newBlockHash);
+}
+
+bool CChainLocksHandler::InternalHasChainLock(const CBlockIndex* pindexPrev, const uint256& newBlockHash)
+{
+    AssertLockHeld(cs_main);
```

Comment:
why do we need cs_main here?

Reply:
Hmm true, not needed. This was a leftover of another version of this method that accessed mapBlockIndex/chainActive.